### ISSUE TITLE
One confusion about the LOMO paper

### ISSUE THREAD

**Issue by Pairshoe** [OpenLMLab/LOMO#79]
Thanks for your great works!

After reading the paper "FULL PARAMETER FINE-TUNING FOR LARGE LANGUAGE MODELS WITH LIMITED RESOURCES", I feel a little bit uncertain about the principal contribution of LOMO.

From my understanding, the standard process of model training is as follows:
1. forward propagation while storing intermediate results, i.e. activations;
2. backward propagation using the intermediate results to calculate gradients and then releasing used intermediate results;
3. calculating the optimizer states, updating the parameters, and then releasing used gradients.

Is the contribution of LOMO that it overlaps the computation of gradients with that of optimizer states? Specially, the gradients per layer are calculated, utilized for optimizer state calculation, and then released immediately?

If this is correct, it seems this technique not restricted solely to the choice of optimizer, and could potentially be employed outside of SGD scope? The choice of SGD in paper is primarily because it does not necessitate optimizer states?

I eagerly anticipate your perceptive response on this matter.

**Comment by KaiLv69** [comment]
Hi, thanks for you interest.

Your understanding is correct. LOMO overlaps the calculations of steps 2 and 3 to save memory. It can indeed be extended to optimizers beyond SGD, such as AdaLomo, for which we have included the code in this repository.
